### DISCUSSION TITLE
Different editing workflows

### DISCUSSION THREAD

**Discussion by kmcnaught** [RaspberryPiFoundation/blockly-keyboard-experimentation#64] [General]
Hi, I couldn't find another place where this was discussed at a high level, please redirect me if there are any existing discussions. 

In my opinion, what I would expect as a user is for these flows to work:

1 - "I want to add a particular thing to my code"
- Go to toolbox
- Find block
- Press something to select it
- Be prompted that it's been copied and I can go somewhere and paste it
- Navigate to where I want it
- Paste it, and focus remains here

2 - "I want to edit this bit of code and add something":
- Navigate to where I want it inserted
- Press something to say I want to insert something
- Be taken to the toolbox to choose what I want
- Press something to select the block
- Have the block inserted, and focus move back where I was

3 - "I want to move / duplicate some existing code"
- Navigate in the code
- Optionally select multiple contiguous blocks
- Perform a cut / copy action
- Navigate somewhere else
- Perform a paste action

Does this match with the current design discussions? Are there other editing flows that we are expecting, that I've missed out?

In particular, I'm finding the "mark" confusing, and if we only need these 2 flows then I would suggest the mark should only be transient while you are in the middle of operation (1). Are there other expected flows where the mark is required? Would we expect anyone to do operation (3) in reverse? i.e. mark somewhere and then go find the code to paste there? Are there other applications with a UX flow similar to this, that we could compare with? I know that some screen readers allow users to set bookmarks/placeholders, but as far as I understand it, that's more for navigation. I'm not aware of a common pattern for this kind of editing / copy/pasting.

**Comment by rachel-fenichel** [comment]
I think you're correct about the main flows. I'm going to add a few others as separate posts for conversation. Note that we do not need to support all of these workflows at the same level, and should probably prioritize ones that are easy to learn even if they are more cumbersome for complex operations.

**Reply by rachel-fenichel** [reply]
I thought we got the idea of the mark from [Milne and Ladner's paper on Blocks4All](https://dl.acm.org/doi/pdf/10.1145/3173574.3173643). But a reread says that they got it from our older accessible Blockly demo, so the actual origin is lost to time.

They describe three editing modes: 
a. Audio-guided drag and drop
b. Location-first select, select, drop (similar to our current marking system)
c. Blocks-first select, select, drop (similar to your option 1, which students preferred for insertion)

> In particular, I'm finding the "mark" confusing, and if we only need these 2 flows then I would suggest the mark should only be transient while you are in the middle of operation (1)

@kmcnaught Based on your logic and the findings in the paper, I agree that we should make the mark transient and otherwise deemphasize it.

> Are there other applications with a UX flow similar to this, that we could compare with?

How about editing an essay, such as when a teacher suggests moving a sentence or a paragraph higher up in the essay? Is this just normally done with copy/paste? Visually I often do it by selecting and dragging.

**Reply by kmcnaught** [reply]
Thanks, I have not read that paper so I will go away and do so. 

I think that moving a sentence or paragraph (or blocks) higher up with a drag operation is still a "source -> destination" flow, there's not an equivalent of the "mark" there. In source code I would either copy/paste in this scenario, or select (possibly multiple) lines and move up with a shortcut like Alt+Up arrow.

**Comment by rachel-fenichel** [comment]
4 - "I want to see what my code does without this (line/chunk)"
- Navigate in the code
- Optionally select multiple contiguous blocks
- Disconnect the blocks, leaving them on the workspace
- Run code
- Optionally undo change

This is relevant for systems that disable orphan blocks, as it lets a user stash constructed code rather than delete it entirely.

Other ways to accomplish something similar:
- Scratch-style backpack
- Trashcan with memory, and train users to look in the trashcan to get their old block stacks out. This has the advantage of being linearly organized.

**Reply by kmcnaught** [reply]
yes, this seems reasonable

**Comment by rachel-fenichel** [comment]
5 - "I want to be able to jump back to the same line of code repeatedly, from anywhere in the workspace"
- Navigate in the code
- Mark a spot
- Continue navigating to read or debug code
- Jump back to mark

My use case here is to read a chunk of code repeatedly to understand or debug it. This probably doesn't need a specific workflow with marks if navigation is easy. For many cases jumping to the root of the stack will accomplish the same thing.

For a complex workspace, this is *maybe* useful--e.g. if a user wants to cross-reference between two block stacks that send and receive messages to each other. I don't know if that's a compelling use case.

**Reply by kmcnaught** [reply]
I think that this should be doable with a screenreader's own functionality, assuming (eventually) we have all the necessary semantics. But I don't know how commonly used this kind of functionality is, particularly for younger /less experienced screen reader users.

Either way, I think it could definitely have use for sighted users too. In a code editor you'd often have this kind of "bookmarking" functionality, with shortcuts to toggle a bookmark for the current line, or navigate next/previous bookmark.
